### PR TITLE
feat: New bloom planning using chunk size TSDB stats

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:

This PR adds a new planning strategy to the bloom builder: `split_by_chunk_size`.
This strategy build tasks looking at the TSDB stats. We add a new configurable `bloom_task_target_chunk_size` where we configure the target chunks size of each task. We keep adding series to a task until the sum of data worth of chunks exceeds the target size. 

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: docs/sources/shared/configuration.md [hunk 1]
```diff
@@ -3775,6 +3775,10 @@ shard_streams:
 # CLI flag: -bloom-build.split-keyspace-by
 [bloom_split_series_keyspace_by: <int> | default = 256]
 
+# Experimental. Target chunk size in bytes for bloom tasks. Default is 100GB.
+# CLI flag: -bloom-build.target-chunk-size
+[bloom_task_target_chunk_size: <int> | default = 100GB]
```
Comment (salvacorts) [inline docs/sources/shared/configuration.md:3780]: Not sure what a good default number would be here. Maybe 100GB is too much. What about 20GB? Wdyt?

File: docs/sources/shared/configuration.md [hunk 2]
```diff
@@ -3765,7 +3765,7 @@ shard_streams:
 [bloom_creation_enabled: <boolean> | default = false]
 
 # Experimental. Bloom planning strategy to use in bloom creation. Can be one of:
-# 'split_keyspace_by_factor'
+# 'split_keyspace_by_factor', 'split_by_chunk_size'
```
Comment (salvacorts) [inline docs/sources/shared/configuration.md:3768]: I'm not entirely happy with this name. Some other options I thought were:
- `split_by_tsdb_chunks_size_stats`
   - This is probably my favorite 
- `split_by_chunks_size_stats`
- `split_by_series_chunks_size`
- `split_by_series_size`

Preferences? Ideas?

Comment (rfratto) [reply docs/sources/shared/configuration.md:3768]: I like `split_by_tsdb_chunks_size` to make it clear which chunks we're talking about; `split_by_tsdb_chunks_size_stats` is fine too, though I'm not sure if the `_stats` bit adds any more information to make it more clear.

File: docs/sources/shared/configuration.md (hunk 1 above)
Comment (rfratto) [reply docs/sources/shared/configuration.md:3780]: What's the impact of setting this too low/too high?

File: docs/sources/shared/configuration.md (hunk 2 above)
Comment (chaudum) [reply docs/sources/shared/configuration.md:3768]: I would prefer `split_by_series_size` or `split_by_stream_size`

Comment (salvacorts) [reply docs/sources/shared/configuration.md:3768]: >I would prefer split_by_series_size or split_by_stream_size

I would like to specifically have "chunks" somewhere in the name provided that in the future we may look at other stats besides the chunk sizing. But I may be overthinking it 😅

File: pkg/bloombuild/planner/strategies/chunksize.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sized series iter: %w", err)
+	}
+
+	tasks := make([]*protos.Task, 0, iterSize)
+	for sizedIter.Next() {
+		series := sizedIter.At()
+		if series.Len() == 0 {
+			// This should never happen, but just in case.
+			level.Error(s.logger).Log("msg", "got empty series batch")
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/bloombuild/planner/strategies/chunksize.go:77]: changed to a warn. 

> should we include extra identifying information about the series
We have no series at this level but the TSDB name. Added it to the log line.

File: pkg/bloombuild/planner/strategies/chunksize.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+		return nil, fmt.Errorf("failed to dedupe blocks: %w", err)
+	}
+
+	return deduped, nil
+}
+
+type seriesWithChunks struct {
+	TSDB   tsdb.SingleTenantTSDBIdentifier
+	FP     model.Fingerprint
+	Chunks []index.ChunkMeta
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/bloombuild/planner/strategies/chunksize.go:157]: Done

File: docs/sources/shared/configuration.md (hunk 2 above)
Comment (salvacorts) [reply docs/sources/shared/configuration.md:3768]: I'm going with `split_by_series_chunks_size`

File: docs/sources/shared/configuration.md (hunk 1 above)
Comment (rfratto) [reply docs/sources/shared/configuration.md:3780]: But it wouldn't change the size of bloom blocks? If that's the case, overpartioning seems better than underpartitioning; 20GB seems fine and we can continue to adjust over time.

**Review by chaudum — APPROVED** [review]

**Pull request merged**